File: Javascript/src/01.AlgorithmicToolbox/Week4DivideAndConquer/majorityElement.js
/**
 * Problem 2 - Majority Element
 * Given a sequence of elements (non-negative integers), check whether it contains an element that appears more than n/2 times.
 * If found, output that element, otherwise, output -1
 * @param {Array.<number>} elements sequence of non-negative integer elements
 * @returns {number} majority element, or -1 if no such element exists found
 */
export function majorityElement(elements) {
  return majorityElementHelper(elements, 0, elements.length);
}

function majorityElementHelper(elements, startIndex, endIndex) {
  if (endIndex - startIndex === 1) {
    return elements[startIndex];
  }
  const middleIndex = Math.floor((endIndex - startIndex) / 2) + startIndex;

  const leftMajority = majorityElementHelper(elements, startIndex, middleIndex);
  const rightMajority = majorityElementHelper(elements, middleIndex, endIndex);
  const majorityThreshold = (endIndex - startIndex) / 2;

  // if not same majority, count
  if (leftMajority !== rightMajority) {
    const leftMajorityCount = getElementCount(
      elements,
      startIndex,
      endIndex,
      leftMajority
    );
    const rightMajorityCount = getElementCount(
      elements,
      startIndex,
      endIndex,
      rightMajority
    );
    if (leftMajorityCount > majorityThreshold) {
      return leftMajority;
    }
    if (rightMajorityCount > majorityThreshold) {
      return rightMajority;
    }
    return -1; // no majority
  }
  // return majority
  return leftMajority;
}

function getElementCount(elements, startIndex, endIndex, elementToCount) {
  let count = 0;
  for (let i = startIndex; i < endIndex; ++i) {
    if (elements[i] === elementToCount) {
      count += 1;
    }
  }
  return count;
}

File: Rust/src/_01_algorithmic_toolbox/week4_divide_and_conquer/majority_element.rs
/// Problem 2 - Majority Element
///
/// Given a sequence of elements (non-negative integers), check whether it contains an element that appears more than n/2 times.
/// If found, output that element, otherwise, output None
pub fn majority_element(elements: Vec<u64>) -> Option<u64> {
    majority_element_helper(&elements[..])
}

fn majority_element_helper(elements: &[u64]) -> Option<u64> {
    let elements_length = elements.iter().count();
    if elements_length == 1 {
        return Some(elements[0]);
    }
    let middle_index = elements_length / 2;

    let left_majority = majority_element_helper(&elements[..middle_index]);
    let right_majority = majority_element_helper(&elements[middle_index..]);

    let majority_threshold = elements_length / 2;

    match (left_majority, right_majority) {
        (Some(left), Some(right)) if left == right => Some(left),
        (Some(left), _) if get_element_count(elements, left) > majority_threshold => Some(left),
        (_, Some(right)) if get_element_count(elements, right) > majority_threshold => Some(right),
        _ => None,
    }
}

fn get_element_count(elements: &[u64], element: u64) -> usize {
    elements.iter().filter(|&x| *x == element).count()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn returns_majority_element_if_exists() {
        assert_eq!(majority_element(vec![2, 3, 9, 2, 2]), Some(2));
        assert_eq!(majority_element(vec![2, 2, 3, 3, 4]), None);
        assert_eq!(majority_element(vec![2, 2, 3, 3]), None);
        assert_eq!(majority_element(vec![3, 3, 2, 3]), Some(3));
        assert_eq!(majority_element(vec![1, 2]), None);
        assert_eq!(majority_element(vec![1, 1]), Some(1));
        assert_eq!(majority_element(vec![1, 2, 3, 4, 5]), None);
    }
}
